### PR TITLE
[SYCL] Implement sycl_ext_oneapi_auto_local_range

### PR DESCRIPTION
**Pull request by 0x12CC**
Add the `auto_range` definition specified in the sycl_ext_oneapi_auto_local_range extension. The runtime already uses `piKernelGetGroupInfo` to determine the local size when it's not specified, so no other changes are needed.

### CODE REVIEW

File: sycl/include/sycl/ext/oneapi/experimental/auto_local_range.hpp [hunk 1]
```diff
@@ -0,0 +1,31 @@
+//==--- auto_local_range.hpp --- SYCL extension for root groups ------------==//
```
Comment (Pennycook) [inline sycl/include/sycl/ext/oneapi/experimental/auto_local_range.hpp:1]:
```suggestion
//==--- auto_local_range.hpp --- SYCL extension for auto range -------------==//
```

**Review by Pennycook — APPROVED** [review]
Just spotted one typo in the header.  Otherwise LGTM!

File: sycl/include/sycl/ext/oneapi/experimental/auto_local_range.hpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+template <> static const inline range<3> auto_range<3>{0, 0, 0};
+
+template <> static const inline range<2> auto_range<2>{0, 0};
+
+template <> static const inline range<1> auto_range<1>{0};
+
+} // namespace experimental
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
```
Comment (steffenlarsen) [inline sycl/include/sycl/ext/oneapi/experimental/auto_local_range.hpp:31]: I believe we should have the `inline namespace _V1` too. And while we are at it we might as well merge some of the namespace declarations..

```suggestion
namespace sycl {
inline namespace _V1 {
namespace ext::oneapi::experimental {

template <int Dimensions> static const inline range<Dimensions> auto_range;

template <> static const inline range<3> auto_range<3>{0, 0, 0};

template <> static const inline range<2> auto_range<2>{0, 0};

template <> static const inline range<1> auto_range<1>{0};

} // ext::oneapi::experimental
} // inline namespace _V1
} // namespace sycl
```

File: sycl/test-e2e/Basic/auto_local_range.cpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
+// RUN: %{build} -I . -o %t.out
+// RUN: %{run} %t.out
+
+#include <cstdlib>
+#include <sycl/ext/oneapi/experimental/auto_local_range.hpp>
```
Comment (steffenlarsen) [inline sycl/test-e2e/Basic/auto_local_range.cpp:5]: Extensions (including experimental ones) are normally included as part of sycl/include/sycl/sycl.hpp.

**Review by steffenlarsen — COMMENTED** [review]
I assume the promotion of the [extension](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_auto_local_range.asciidoc) will happen in a follow-up patch? When you do, please remember to define `SYCL_EXT_ONEAPI_AUTO_LOCAL_RANGE` in sycl/source/feature_test.hpp.in. Alternatively you can define the feature test macro as part of this patch.

**Comment by Pennycook** [comment]
> @Pennycook, I discovered a problem with this proposal while making the other changes. Presently, including `<sycl/sycl.hpp>` doesn't perform any global construction. I don't think it's possible to implement your proposal without using global constructors. I've made some minor changes to your extension to make `auto_range` a function and resolve this issue.

Oh, that's annoying. I definitely prefer the version where this is a variable instead of a function.

Just to double-check I understand: if this could be declared `constexpr` then we wouldn't have any problems, right?

**Comment by Pennycook** [comment]
> I believe that's correct. Global variables declared `constexpr` should have static initialization so they won't have any of the issues related to global constructors. I've updated the note in your extension to highlight this difference.

Thanks.  I'm happy for this to proceed as-is, and we can revisit later if the `constexpr` situation changes.

**Review by Pennycook — APPROVED** [review]

**Comment by gmlueck** [comment]
If this PR fully implements the extension, it should also move the extension document to the "experimental" directory and update the Status section according to the [template](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/template.asciidoc#status).

File: sycl/include/sycl/ext/oneapi/experimental/auto_local_range.hpp [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/range.hpp>
+
+#define SYCL_EXT_ONEAPI_AUTO_LOCAL_RANGE 1
```
Comment (steffenlarsen) [inline sycl/include/sycl/ext/oneapi/experimental/auto_local_range.hpp:13]: This should be done in sycl/source/feature_test.hpp.in instead.

**Review by steffenlarsen — APPROVED** [review]
LGTM!

**Review by gmlueck — APPROVED** [review]

**Comment by steffenlarsen** [comment]
Windows testing pre-merge only failed with known annotated_arg failures (https://github.com/intel/llvm/issues/11224)

Merging.

**Pull request merged**